kratos/utilities/math_utils.h

```diff
@@ -464,243 +408,239 @@ class MathUtils
+            if(rInvertedMatrix.size1() != size1 || rInvertedMatrix.size2() != size2) {
+                rInvertedMatrix.resize(size1, size2,false);
             }
 
 #ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
-            Matrix temp(InputMatrix);
+            Matrix temp(rInputMatrix);
             AMatrix::LUFactorization<MatrixType, DenseVector<std::size_t> > lu_factorization(temp);
-            InputMatrixDet = lu_factorization.determinant();
-            KRATOS_ERROR_IF(std::abs(InputMatrixDet) <= ZeroTolerance) << "Matrix is singular: " << InputMatrix << std::endl;
-            InvertedMatrix = lu_factorization.inverse();
+            rInputMatrixDet = lu_factorization.determinant();
+            KRATOS_ERROR_IF(std::abs(rInputMatrixDet) <= ZeroTolerance) << "Matrix is singular: " << rInputMatrix << std::endl;
+            rInvertedMatrix = lu_factorization.inverse();
 #else
 
             typedef permutation_matrix<SizeType> pmatrix;
-            Matrix A(InputMatrix);
+            Matrix A(rInputMatrix);
             pmatrix pm(A.size1());
             const int singular = lu_factorize(A,pm);
-            InvertedMatrix.assign( IdentityMatrix(A.size1()));
-            KRATOS_ERROR_IF(singular == 1) << "Matrix is singular: " << InputMatrix << std::endl;
-            lu_substitute(A, pm, InvertedMatrix);
+            rInvertedMatrix.assign( IdentityMatrix(A.size1()));
+            KRATOS_ERROR_IF(singular == 1) << "Matrix is singular: " << rInputMatrix << std::endl;
+            lu_substitute(A, pm, rInvertedMatrix);
 
             // Calculating determinant
-            InputMatrixDet = 1.0;
+            rInputMatrixDet = 1.0;
 
             for (IndexType i = 0; i < A.size1();++i) {
                 IndexType ki = pm[i] == i ? 0 : 1;
-                InputMatrixDet *= (ki == 0) ? A(i,i) : -A(i,i);
+                rInputMatrixDet *= (ki == 0) ? A(i,i) : -A(i,i);
             }
 
  #endif // ifdef KRATOS_USE_AMATRIX
+       } else {
+           KRATOS_ERROR << "Not possible to invert Matrix: " << rInputMatrix << std::endl;
+       }
+
+       // Checking condition number
+       if (Tolerance > 0.0) { // Check is skipped for negative tolerances
+            CheckConditionNumber(rInputMatrix, rInvertedMatrix, Tolerance);
        }
     }
 
     /**
-     * It inverts matrices of order 2 //VERIFIED!!!
-     * @param InputMatrix Is the input matrix (unchanged at output)
-     * @param InvertedMatrix Is the inverse of the input matrix
-     * @param InputMatrixDet Is the determinant of the input matrix
+     * @brief It inverts matrices of order 2
+     * @param rInputMatrix Is the input matrix (unchanged at output)
+     * @param rInvertedMatrix Is the inverse of the input matrix
+     * @param rInputMatrixDet Is the determinant of the input matrix
      */
-
+    template<class TMatrix1, class TMatrix2>
```

Comment:
also here ?

Reply:
Idem, to allow bounded and unbounded

Reply:
Of course can be used with sparse, but a sparse system with 4x4 is quite small